### ISSUE TITLE
New rule proposal: avoid several empty lines in a row.

### ISSUE THREAD

**Issue by niconoe-** [sqlfluff/sqlfluff#2634]
I don't know if several empty lines is something that makes sense in some dialects, but if not, I think SQLFluff should have a rule to flag this.

**Anti-pattern**
```sql
SELECT * FROM my_table;


SELECT * FROM my_other_table;
```

**Best practice**
```sql
SELECT * FROM my_table;

SELECT * FROM my_other_table;
```

Removing all empty lines makes no sense at all as it adds clean space to separate a list of statements, like a new paragraph in a novel. But having too much of them in a row creates a too big space that can make your SQL hard to read.

This style rule is also present in [JavaScript](https://eslint.org/docs/rules/no-multiple-empty-lines) for instance, and probably from other languages.

Technically, I would go on looking for `newline` segment type, and if the (at least) 2 next segments type encountered are also `newline`s, the rule should flag. The fix to provide is to simply keep 2 newlines in a row when 3+ are found.

Thanks a lot!

**Comment by jpy-git** [comment]
yeah this rule is 100% necessary, the suggested approach is sensible 👍

**Comment by jpy-git** [comment]
Thought a bit more on this rule and think we need some slight refinement before we implement:

The initial suggested approach makes sense to have TWO newlines BETWEEN statements, however, this rule should also ensure ONE newline WITHIN statements:

**Anti-pattern**
```sql
select foo

from bar;
```
**Best practice**
```sql
select foo
from bar;
```

The one exception to this being between the final CTE and the start of a statement which should have 2 newlines according to L022. I suggest we replace L022 with this rule and have the CTE spacing be an option to this rule.

There should also be some consideration for comments in between lines. To keep things simple and consistent, I would suggest:
- Within statements: One newline before and after comment
**Anti-pattern**
```sql
select foo

--some comment

from bar;
```
**Best practice**
```sql
select foo
--some comment
from bar;
```
- Between statements: Two newlines before and after comment
**Anti-pattern**
```sql
select foo
from bar;


--some comment
select foo
from bar;
```
**Best practice**
```sql
select foo
from bar;

--some comment

select foo
from bar;
```

@barrywhart @tunetheweb sound reasonable?

**Comment by barrywhart** [comment]
@jpy-git: I'll take a look at your notes later when I have some down time.

Should we consider making this part of rule L052? Currently, that rule adds a semicolon at the end of the **file**, but there was a request (and my initial, close but buggy attempt) to extend it to "semicolon after each statement". It seems like it might be worth considering to have a single rule enforcing both the semicolon and the line spacing.

I made some notes on the other issue about this. It would be good to simplify L052 if possible, perhaps using `awesome-pattern-matching`. The number of tricky `if` statements in the current implementation made it tricky to extend (for me, anyway).

**Comment by niconoe-** [comment]
I really like the idea of also limiting the number of newlines to only 1 WITHIN statements. 

I took a look at L052 and I found this rule is really oriented to the semi-colon placement (that should be rather called « delimiter placement » as the delimiter of statements can be changed by anything else for instance in MySQL, but that’s another topic).
IMO, my proposal is not linked to the delimiters even if it’s linked to statements. In fact, including this proposal into L052 could be very confusing because considering the following example, there is not a single semi-colon and in my mind, this new rule should flag:

**Anti-pattern**

```sql
— This is a comment


— This is another comment
```

**Best pattern**

```sql
— This is a comment

— This is another comment
```

Actually, I think that giving L052 the additional responsibility to manage the new lines could make it too complex.

Thank you for considering my proposal 😄

**Comment by jpy-git** [comment]
I think we should keep this separate from L052. L052 just moves/inserts semicolons (the proposed bit to add to the end of all segments is T-SQL specific, and the T-SQL approach to semicolons is pretty messy in the grammar atm) so lets not overcomplicate further.

This rule would just focus on newline spacing. So this plus L009 and L050 would fully cover vertical alignment in the file. Potentially we can combine those in the future but initially I think this can be it's own rule.

**Comment by dennysrega1ado** [comment]
+1

**Comment by CarbonChauvinist** [comment]
Can this rule be leveraged in some way to allow for _the addition of blank new lines within statements_ (primarily JOIN clauses)? For instance, sometimes especially when there are a ton of joins it easier to grok the following ...

```
...
INNER JOIN Table1 AS t1
    ON t1.col = e1.col

INNER JOIN Table2 AS t2
    ON t1.col = t2.col

INNER JOIN Table3 AS t3
    ON t2.col = t3.col
...
````

than the following:
```
...
INNER JOIN Table1 AS t1
    ON t1.col = e1.col
INNER JOIN Table2 AS t2
    ON t1.col = t2.col
INNER JOIN Table3 AS t3
    ON t2.col = t3.col
...
````
